src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -1325,40 +1335,161 @@ private function generateOverriddenGetters($id, Definition $definition, \Reflect
+            $partial[] = 'switch (func_num_args()) {';
+
+            if ($numReqArgs = key($optionalArgs)) {
+                for ($i = 0; $i < $numReqArgs; ++$i) {
+                    $partial[] = "case $i:";
+                }
+            }
+
+            foreach ($optionalArgs as $i => $value) {
+                if (null === $value && !$params[$i]->allowsNull()) {
+                    throw new RuntimeException(sprintf('Unable to configure service "%s": argument %d ($%s) of partial method "%s::%s()" must be non-null.', $id, $i, $params[$i]->name, $class->name, $r->name));
+                }
+                $partial[] = "case $i: \${$params[$i]->name} = \$c{$salt}->resolveServices(\$b{$salt}->unescapeValue(\$b{$salt}->resolveValue(\$v{$salt}[$i])));";
+            }
+
+            $partial[] = sprintf("}\nreturn parent::%s;\n}", $call);
+            $partials .= implode("\n", $partial);
+        }
+
+        return $partials;
+    }
+
     /**
      * @internal
      */
-    public static function getGetterReflector(\ReflectionClass $class, $name, $id, &$type)
+    public static function getPartialReflector(\ReflectionClass $class, $name, $id, &$type, $msg = 'partial method')
     {
         if (!$class->hasMethod($name)) {
-            throw new RuntimeException(sprintf('Unable to configure getter injection for service "%s": method "%s::%s" does not exist.', $id, $class->name, $name));
+            throw new RuntimeException(sprintf('Unable to configure %s for service "%s": method "%s::%s" does not exist.', $msg, $id, $class->name, $name));
         }
         $r = $class->getMethod($name);
         if ($r->isPrivate()) {
-            throw new RuntimeException(sprintf('Unable to configure getter injection for service "%s": method "%s::%s" must be public or protected.', $id, $class->name, $r->name));
+            throw new RuntimeException(sprintf('Unable to configure %s for service "%s": method "%s::%s" must be public or protected.', $msg, $id, $class->name, $r->name));
         }
         if ($r->isStatic()) {
-            throw new RuntimeException(sprintf('Unable to configure getter injection for service "%s": method "%s::%s" cannot be static.', $id, $class->name, $r->name));
+            throw new RuntimeException(sprintf('Unable to configure %s for service "%s": method "%s::%s" cannot be static.', $msg, $id, $class->name, $r->name));
         }
         if ($r->isFinal()) {
-            throw new RuntimeException(sprintf('Unable to configure getter injection for service "%s": method "%s::%s" cannot be marked as final.', $id, $class->name, $r->name));
+            throw new RuntimeException(sprintf('Unable to configure %s for service "%s": method "%s::%s" cannot be marked as final.', $msg, $id, $class->name, $r->name));
         }
+        if ($type = method_exists($r, 'getReturnType') ? $r->getReturnType() : null) {
+            $type = ': '.($type->allowsNull() ? '?' : '').self::generateTypeHint($type, $r);
```

Comment:
I vote for moving the building of the dumped return type outside this method. I already found the reference argument harder to read previously, but it becomes worse when nesting the resolution by reference in another method. We already return all the necessary info anyway, as this logic depends only on the reflector anyway